### PR TITLE
GOSSIP-41 Transfer gossip data in bulk

### PR DESCRIPTION
**Pull request by pxsalehi**
This is the first version I had in mind. The simplest way is to create a bulk message which carries a list of messages. The handler simply unpacks the list of messages and the rest will stay the same. I have failing tests to go through yet. But maybe someone can let me know if this is the way to go or there is a better alternative.

### CODE REVIEW

File: gossip-base/src/main/java/org/apache/gossip/model/PerNodeDataBulkMessage.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package org.apache.gossip.model;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/model/PerNodeDataBulkMessage.java]: No author comments.

File: gossip-base/src/main/java/org/apache/gossip/model/PerNodeDataBulkMessage.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package org.apache.gossip.model;
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/model/PerNodeDataBulkMessage.java:18]: All files Requires apache licence header

File: gossip-base/src/main/java/org/apache/gossip/model/PerNodeDataBulkMessage.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * Created by pxsalehi on 22.08.17.
+ */
+public class PerNodeDataBulkMessage extends Base {
+    private List<PerNodeDataMessage> messages = new ArrayList<>();
+
+    public void addMessage(PerNodeDataMessage msg) {
+        messages.add(msg);
+    }
+
+    public List<PerNodeDataMessage> getMessages() {
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/model/PerNodeDataBulkMessage.java]: Indent is two spaces. Use our style sheet.

File: gossip-base/src/main/java/org/apache/gossip/manager/handlers/MessageHandlerFactory.java [hunk 4]
```diff
@@ -35,8 +30,8 @@ public static MessageHandler defaultHandler() {
     return concurrentHandler(
         new TypedMessageHandler(Response.class, new ResponseHandler()),
         new TypedMessageHandler(ShutdownMessage.class, new ShutdownMessageHandler()),
-        new TypedMessageHandler(PerNodeDataMessage.class, new PerNodeDataMessageHandler()),
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/manager/handlers/MessageHandlerFactory.java]: Lets not remove the old system. Eventually I think the bulk only way makes sense but there is no issue in supporting both in the code and choosing with a switch

File: gossip-base/src/main/java/org/apache/gossip/manager/AbstractActiveGossiper.java [hunk 5]
```diff
@@ -109,22 +120,31 @@ public final void sendPerNodeData(LocalMember me, LocalMember member){
...
+        PerNodeDataMessage message = new PerNodeDataMessage();
         message.setExpireAt(innerEntry.getValue().getExpireAt());
         message.setKey(innerEntry.getValue().getKey());
         message.setNodeId(innerEntry.getValue().getNodeId());
         message.setTimestamp(innerEntry.getValue().getTimestamp());
         message.setPayload(innerEntry.getValue().getPayload());
         message.setReplicable(innerEntry.getValue().getReplicable());
-        gossipCore.sendOneWay(message, member.getUri());   
+        udpMessage.addMessage(message);
+        if (udpMessage.getMessages().size() == 100) {
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/manager/AbstractActiveGossiper.java]: This needs to be a variable that can be controlled through settings

File: gossip-base/src/main/java/org/apache/gossip/manager/AbstractActiveGossiper.java [hunk 6]
```diff
@@ -84,22 +84,33 @@ public final void sendSharedData(LocalMember me, LocalMember member){
...
+      SharedDataMessage message = new SharedDataMessage();
       message.setExpireAt(innerEntry.getValue().getExpireAt());
       message.setKey(innerEntry.getValue().getKey());
       message.setNodeId(innerEntry.getValue().getNodeId());
       message.setTimestamp(innerEntry.getValue().getTimestamp());
       message.setPayload(innerEntry.getValue().getPayload());
       message.setReplicable(innerEntry.getValue().getReplicable());
-      gossipCore.sendOneWay(message, member.getUri());
+      udpMessage.addMessage(message);
+      if (udpMessage.getMessages().size() == 100) {
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/manager/AbstractActiveGossiper.java]: Use the same variable to control this.

**Review by edwardcapriolo — CHANGES_REQUESTED** [review]

**Comment by pxsalehi** [comment]
I still have to add tests for bulk transfer. But maybe you can check it out and let me know if there are any other issues.

**Comment by pxsalehi** [comment]
Instead of new tests I have used parameterized tests to run existing tests with bulk transfer activated.

File: gossip-base/src/main/java/org/apache/gossip/manager/AbstractActiveGossiper.java [hunk 7]
```diff
@@ -78,12 +79,21 @@ public final void sendShutdownMessage(LocalMember me, LocalMember target){
...
-    if (member == null){
+
+  public final void sendSharedData(LocalMember me, LocalMember member) {
+    if (member == null) {
       return;
     }
     long startTime = System.currentTimeMillis();
+    if (gossipSettings.isBulkTransfer()) {
+      sendSharedDataInBulkInternal(me, member);
+    } else
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/manager/AbstractActiveGossiper.java]: use { } even for one line if so your intention is clear.

File: gossip-base/src/main/java/org/apache/gossip/manager/AbstractActiveGossiper.java [hunk 8]
```diff
@@ -92,22 +102,59 @@ public final void sendSharedData(LocalMember me, LocalMember member){
...
+  }
+
   public final void sendPerNodeData(LocalMember me, LocalMember member){
     if (member == null){
       return;
     }
     long startTime = System.currentTimeMillis();
+    if (gossipSettings.isBulkTransfer()) {
+      sendPerNodeDataInBulkInternal(me, member);
+    } else
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/manager/AbstractActiveGossiper.java]: same deal use { } even for one line if

File: gossip-itest/src/test/java/org/apache/gossip/PerNodeDataEventTest.java [hunk 9]
```diff
@@ -23,45 +23,57 @@
...
-  
-  
+  private int base;
+  private boolean bulkTransfer;
+
+  public PerNodeDataEventTest(int base, boolean bulkTransfer) {
+    this.base = base;
+    this.bulkTransfer = bulkTransfer;
+  }
+
```
Comment (edwardcapriolo) [inline gossip-itest/src/test/java/org/apache/gossip/PerNodeDataEventTest.java:51]: Slick. I like this.

**Review by edwardcapriolo — CHANGES_REQUESTED** [review]
Looks good. I left a comment with minor changes.

File: gossip-itest/src/test/java/org/apache/gossip/PerNodeDataEventTest.java (hunk 9 above)
Comment (pxsalehi) [reply gossip-itest/src/test/java/org/apache/gossip/PerNodeDataEventTest.java:51]: While going through the tests, I noticed the same can be done in `TenNodeThreeSeedTest` by parameterizing base.

Comment (edwardcapriolo) [reply gossip-itest/src/test/java/org/apache/gossip/PerNodeDataEventTest.java:51]: Lets handle that in a separate ticket.

**Comment by pxsalehi** [comment]
Done!

**Pull request merged**